ls1intum/bamboo-server-notification-plugin
# Feature/send artifacts

### Checklist


### Motivation and Context

As the first step for integrating static code analysis into programming exercises, I implemented the functionality to send parsed static code analysis reports as part of notifications to Artemis. Currently only spotbugs reports are supported.

In order to use the new functionality, test repositories and build plans must be manually configured:

**Repository**
Test repositories will contain configurations for static code analysis tools. Java programming exercises will use maven plugins of static code analysis tools to perform the analysis.
Example test repository configuring spotbugs and checkstyle: [test repository](https://bitbucket.ase.in.tum.de/projects/STCSCATN1/repos/stcscatn1-tests/browse)

**Build Plan**
An additional task, which executes static code analysis maven goals, must be created.
To keep the produced reports, artifact definitions must be created. For each static code analysis tool an artifact definition will be created matching the filename of the generated report.
Example BASE plan with analysis task and artifact definitions: [base plan](https://bamboo.ase.in.tum.de/browse/STCSCATN1-BASE)

### Description

**Format**
JSON objects with the following format will be created:
`{"tool":"spotbugs","issues":[{"classname":"com.stefan.staticCodeAnalysis.App","type":"ES_COMPARING_PARAMETER_STRING_WITH_EQ","priority":"High","category":"BAD_PRACTICE","message":"Comparison of String parameter using == or != in com.stefan.staticCodeAnalysis.App.equalString(String)","line":16}]}`
The format will be extended, if other tools provide more information.

**Error handling**
If artifact handling or report parsing fails, a notification containing the test results will still be sent.

### Steps for Testing

Testing will ensure, that the existing programming exercise functionality still works, as Artemis can't utilize the new incoming data at the moment.
Isolated testing on a test server is not possible, because we only have one Bamboo instance.

**Install the new version of the plugin locally:**

1. Checkout this branch and compile the plugin yourself following the readme description or download the plugin here
2. Log in to Bamboo
3. Select manage apps
![manageapps](https://user-images.githubusercontent.com/16407766/81596941-2cf02f00-93c5-11ea-86ac-10690733309e.PNG)

4. Install the bamboo plugin
![uploadPlugin](https://user-images.githubusercontent.com/16407766/81597013-4b562a80-93c5-11ea-9ef3-7f61115ff8bb.PNG)

**Local testing:**
Checkout Artemis develop branch. **Please test with all roles (admin, instructor, tutor, student)**

**Test without static code analysis configured:
1. Log in to Artemis
2. Navigate to Course Management
3. Create a programming exercise
4. Navigate to the course management exercise details page by clicking on the highlighted link below
![courseManExDet](https://user-images.githubusercontent.com/16407766/81602605-31b8e100-93cd-11ea-800d-24bc4a65649a.PNG)

5. Inspect the template and solution results
![results](https://user-images.githubusercontent.com/16407766/81602831-96743b80-93cd-11ea-9db3-bba1c87a04e4.PNG)

=> Solution result should be 100% and feedback should be available for the template result


**Test with static code analysis configured:
1. Log in to Artemis
2. Navigate to Course Management
3. Create a Java programming exercise
4. Starting at the properties tag, replace the content of the test repository pom.xml with the respective content of the pom.xml found in the [example test repository](https://bitbucket.ase.in.tum.de/projects/STCSCATN1/repos/stcscatn1-tests/browse)
5. Update the BASE and SOLUTION build plan using the [sample plan](https://bamboo.ase.in.tum.de/browse/STCSCATN1-BASE) as a template. Create a new task which runs the static code analysis and create artifact definitions. The picture below shows, where artifact definitions can be created. 
![creating-artifact-definition-in-Bamboo](https://user-images.githubusercontent.com/16407766/82714566-8858d180-9c8f-11ea-9d55-19508c1e9655.png)


6. Navigate to the course management exercise details page
7. Resubmit the template and solution repository by clicking on the blue buttons as seen below
![resubmit](https://user-images.githubusercontent.com/16407766/81621051-b7e51f80-93ed-11ea-84cd-1437b87e6376.PNG)

=> Solution result should be 100% and feedback should be available for the template result
